### PR TITLE
Write log entries as json

### PR DESCRIPTION
**Pull request by dgageot**
This is the first step in supporting structured success/error data in `memlogd` log files.

Those log files are currently made of semicolon separated fields and, for each line, the list of fields is the same.

We would like to support multiple kinds of entries. Application logs, like we do already, but also structured success/error data. Those entries would have some fields in common and some based on the type of entry. 

Switching from semicolon separated fields to json gives us this flexibility without much added complexity. In fact, it even removes a constraint we currently have on `Source` field not containing a semicolon.

Signed-off-by: David Gageot <david.gageot@docker.com>

### CODE REVIEW

File: pkg/memlogd/cmd/logread/main.go [hunk 1]
```diff
@@ -49,7 +61,13 @@ func main() {
 		panic(err)
 	}
 
-	r := bufio.NewReader(conn)
-	r.WriteTo(os.Stdout)
+	var entry LogEntry
+	decoder := json.NewDecoder(conn)
+	for {
+		if err := decoder.Decode(&entry); err != nil {
+			panic(err)
```
Comment (djs55) [inline pkg/memlogd/cmd/logread/main.go:68]: I almost panicked at the `panic` but then I saw the `panic` above: I guess this is how we handle errors in this utility (which is fine)

Comment (dgageot) [reply pkg/memlogd/cmd/logread/main.go:68]: Yes...

File: pkg/memlogd/cmd/memlogd/main.go [hunk 2]
```diff
@@ -213,27 +217,20 @@ func readLogFromFd(maxLineLen int, fd int, source string, logCh chan logEntry) {
...
-				if err := syscall.Close(msg.fd); err != nil {
-					doLog(logCh, fmt.Sprintf("ERROR: failed to close fd: %s", err))
-				}
-				continue
-			}
-			go readLogFromFd(lineMaxLength, msg.fd, msg.name, logCh)
-		}
+func loggingRequestHandler(lineMaxLength int, logCh chan LogEntry, fdMsgChan chan fdMessage) {
+	for msg := range fdMsgChan {
+		go readLogFromFd(lineMaxLength, msg.fd, msg.name, logCh)
```
Comment (djs55) [inline pkg/memlogd/cmd/memlogd/main.go:233]: This looks more concise!

File: pkg/memlogd/cmd/memlogd/main_test.go [hunk 3]
```diff
@@ -113,63 +112,6 @@ func TestFinite2(t *testing.T) {
 	}
 }
 
-func TestGoodName(t *testing.T) {
-	// Test that the source names can't contain ";"
```
Comment (djs55) [inline pkg/memlogd/cmd/memlogd/main_test.go:117]: Ah it's a nice feature that we don't need to test this any more.

**Review by djs55 — APPROVED** [review]
It looks like a nice improvement to me!

**Comment by deitch** [comment]
Except that `logwrite` wasn't synced with this. Oops. It caused a bunch of failures for CI now.

We need to get logwrite synced.

**Pull request merged**